README
NAME
    Filesys::Notify::KQueue - Wrap IO::KQueue for watching file system.

SYNOPSIS
      use Filesys::Notify::KQueue;

      my $notify = Filesys::Notify::KQueue->new(
          path    => [qw(~/Maildir/new)],
          timeout => 1000,
      );
      $notify->wait(sub {
          my @events = @_;

          foreach my $event (@events) {
              ## ....
          }
      });

DESCRIPTION
    Filesys::Notify::KQueue is IO::KQueue wrapper for watching file system.

METHODS
  new - Hash or HashRef
    This is constructor method.

    path - ArrayRef[Str]
        Watch files or directories.

    timeout - Int
        KQueue's timeout. (milli second)

  wait - CodeRef
    There is no file name based filter. Do it in your own code. You can get
    types of events (create, modify, rename, delete).

AUTHOR
    Kenta Sato <karupa@cpan.org>

SEE ALSO
    IO::KQueue Filesys::Notify::Simple AnyEvent::Filesys::Notify
    File::ChangeNotify Mac::FSEvents Linux::Inotify2

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Filesys::Notify::KQueue

{{$NEXT}}

0.11 2015-06-07T11:29:51Z

    - Fixed dependency. (thanks SREZIC)

0.10 2015-06-06T01:28:24Z

    - Allow to set zero/undef as `timeout`. Timeout implementation doesn't match IO::KQueue. (thanks TLINDEN)

0.09 2013-10-12T02:43:02Z

    - minil migrate
    - fixed pod spelling.
    - fixed dependency.

0.08    Wed Dec 14 2011

    - File::Path after version 2.06_05 required at test.(oops)

0.07    Sun Dec 11 2011

    - add t::Util to MANIFEST.(oops)

0.06    Sun Dec 11 2011

    - [bugfix] cannot set timeout.
    - [bugfix] cannnot get all event if create nested directories.
    - add test "02_nest.t"
    - fixed spell miss.
    - coding style is changed a little.
    - not import File::Find's export functions.

0.05    Thu Oct 20 2011

    - build_require to test_require.
    - fixed a day of the week in Changes.

0.04    Thu Oct 20 2011

    - fixed test dependency. (thanks ANDK)

0.03    Sat Sep 24 2011

    - fixed default timeout.
    - write pod.

0.02    Sun Sep 18 2011

    - wait method is compatible with the interface of Filesys::Notify::Simple now.

0.01    Sun Sep 11 2011

    - original version


t/Util.pm
use strict;
use warnings;
use utf8;

use Carp ();
use IO::Handle;
use Test::More;
use Test::SharedFork;
use File::Find;
use File::Path qw/make_path remove_tree/;
use parent qw/Exporter/;
our @EXPORT_OK = qw/create_paths touch delete_paths cmp_files get_filelist/;
our @EXPORT = qw/test_fork parent child/;

sub import {
    my $class = shift;
    $class->export_to_level(1, @_);
    strict->import;
    warnings->import;
    utf8->import;
    Test::More->import;
}

sub create_paths {
    my @test_paths = @_;

    my @dirs  = sort { length($a =~ m{/}g) < length($b =~ m{/}g) } grep { $_ =~ m{/$} } @test_paths;
    my @files = grep { $_ !~ m{/$} } @test_paths;

    make_path(@dirs, +{
        mode => 0777,
    });
    touch(@files);
}

sub touch {
    my @files = @_;
    foreach my $file (@files) {
        open my $fh, '>', $file or die $!;
        print $fh 'foo';
        close $fh;
    }
}

sub delete_paths {
    my @test_paths = @_;

    my @dirs  = grep { -d $_ } @test_paths;
    my @files = grep { -f $_ } @test_paths;

    unlink(@files);
    remove_tree(@dirs);
}

sub cmp_files {
    local $Test::Builder::Level = $Test::Builder::Level + 1;
    my($files1, $files2) = @_;

    is_deeply [sort {$a cmp $b} @$files1] => [sort {$a cmp $b} @$files2], 'watching all file';
}

sub get_filelist {
    my $path = shift;
    my @files;

    find +{
        wanted => sub { push @files => $File::Find::name },
        no_chdir => 1,
    } => $path;

    return @files;
}

sub parent (&);
sub child (&);
sub test_fork (&) {## no critic
    my $code = shift;

    my $child_code;
    my $parent_code;
    {
        local *child = sub (&) {## no critic
            $child_code = shift;
        };
        local *parent = sub (&) {## no critic
            $parent_code = shift;
        };
        $code->();
    }

    pipe(my $child_rdr,   my $parent_wtr);
    pipe(my $parent_rdr,  my $child_wtr);
    $parent_wtr->autoflush(1);
    $child_wtr->autoflush(1);

    if ($parent_code and $child_code) {
        my $pid = fork;
        if ($pid == 0) {
            Test::SharedFork->child;
            close $parent_rdr;
            close $parent_wtr;
            $child_code->($child_rdr, $child_wtr);
            close $parent_rdr;
            close $parent_wtr;
            exit(0);
        }
        elsif ($pid != 0) {
            Test::SharedFork->parent;
            close $child_rdr;
            close $child_wtr;
            $parent_code->($parent_rdr, $parent_wtr);
            close $parent_rdr;
            close $parent_wtr;
            waitpid $pid, 0;
        }
        else {
            die $!;
        }
    }
    else {
        Carp::croak 'parent code or child code not found.';
    }
}

1;


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Filesys-Notify-KQueue',
    module_name     => 'Filesys::Notify::KQueue',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Filesys/Notify/KQueue.pm
package Filesys::Notify::KQueue;
use strict;
use warnings;
our $VERSION = '0.10';

use File::Find ();
use IO::KQueue;

sub default_timeout { 1000 }

sub new {
    my $class = shift;
    my $args  = (@_ == 1) ? $_[0] : +{ @_ };
    my $self  = bless(+{} => $class);

    $self->timeout(exists $args->{timeout} ? $args->{timeout} : $class->default_timeout);
    $self->{_kqueue} = $args->{kqueue} if exists($args->{kqueue});
    $self->add(@{$args->{path}})       if exists($args->{path});

    return $self;
}

sub kqueue {
    my $self = shift;
    $self->{_kqueue} ||= IO::KQueue->new;
}

sub timeout {
    my $self = shift;
    (@_ == 1) ? ($self->{_timeout} = shift) : $self->{_timeout};
}

sub add {
    my $self = shift;

    foreach my $path (@_) {
        next if exists($self->{_files}{$path});
        if (-f $path) {
            $self->add_file($path);
        }
        elsif (-d $path) {
            $self->add_dir($path);
        }
        else {
            die "Unknown file '$path'";
        }
    }
}

sub add_file {
    my($self, $file) = @_;

    $self->{_files}{$file} = do {
        open(my $fh, '<', $file) or die("Can't open '$file': $!");
        die "Can't get fileno '$file'" unless defined fileno($fh);

        # add to watch
        $self->kqueue->EV_SET(
            fileno($fh),
            EVFILT_VNODE,
            EV_ADD | EV_CLEAR,
            NOTE_DELETE | NOTE_WRITE | NOTE_RENAME | NOTE_REVOKE,
            0,
            $file,
        );

        $fh;
    };
}

sub add_dir {
    my($self, $dir) = @_;

    $self->add_file($dir);
    File::Find::find +{
        wanted => sub { $self->add($File::Find::name) },
        no_chdir => 1,
    } => $dir;
}

sub files        { keys   %{shift->{_files}} }
sub file_handles { values %{shift->{_files}} }
sub get_fh       {
    my %files = %{shift->{_files}};
    @files{@_};
}

sub unwatch {
    my $self = shift;
    my @path = @_;

    foreach my $path (@_) {
        close($self->{_files}{$path});
        delete($self->{_files}{$path});
    }
}

sub wait {
    my ($self, $cb) = @_;

    my $events = $self->get_events;
    if ($self->timeout) {
        until (@$events) {
            $events = $self->get_events;
        }
    }

    $cb->(@$events);
}

sub get_events {
    my $self = shift;

    my @kevents = $self->kqueue->kevent($self->timeout);

    my @events;
    foreach my $kevent (@kevents) {
        my $path  = $kevent->[KQ_UDATA];
        my $flags = $kevent->[KQ_FFLAGS];

        if(($flags & NOTE_DELETE) or ($flags & NOTE_RENAME)) {
            my $event = ($flags & NOTE_DELETE) ? 'delete' : 'rename';

            if (-d $path) {
                my @stored_paths = grep { m{^${path}/} } $self->files;
                $self->unwatch(@stored_paths);
                push @events => map {
                    +{
                        event => $event,
                        path  => $_,
                    }
                } @stored_paths;
            }

            $self->unwatch($path);
            push @events => +{
                event => $event,
                path  => $path,
            };
        }
        elsif ($flags & NOTE_WRITE) {
            if (-f $path) {
                push @events => +{
                    event => 'modify',
                    path  => $path,
                };
            }
            elsif (-d $path) {
                File::Find::finddepth +{
                    wanted => sub {
                        return if exists($self->{_files}{$File::Find::name});
                        push @events => +{
                            event => 'create',
                            path  => $File::Find::name,
                        };
                        $self->add($File::Find::name);
                    },
                    no_chdir => 1,
                } => $path;
            }
        }
    }

    return \@events;
}

1;
__END__

=encoding utf-8

=for stopwords KQueue's

=head1 NAME

Filesys::Notify::KQueue - Wrap IO::KQueue for watching file system.

=head1 SYNOPSIS

  use Filesys::Notify::KQueue;

  my $notify = Filesys::Notify::KQueue->new(
      path    => [qw(~/Maildir/new)],
      timeout => 1000,
  );
  $notify->wait(sub {
      my @events = @_;

      foreach my $event (@events) {
          ## ....
      }
  });

=head1 DESCRIPTION

Filesys::Notify::KQueue is IO::KQueue wrapper for watching file system.

=head1 METHODS

=head2 new - Hash or HashRef

This is constructor method.

=over 4

=item path - ArrayRef[Str]

Watch files or directories.

=item timeout - Int

KQueue's timeout. (millisecond)

=back

=head2 wait - CodeRef

There is no file name based filter. Do it in your own code.
You can get types of events (create, modify, rename, delete).

=head1 AUTHOR

Kenta Sato E<lt>karupa@cpan.orgE<gt>

=head1 SEE ALSO

L<IO::KQueue> L<Filesys::Notify::Simple> L<AnyEvent::Filesys::Notify> L<File::ChangeNotify> L<Mac::FSEvents> L<Linux::Inotify2>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
